core/impl/src/main/java/org/jboss/jca/core/connectionmanager/pool/mcp/SemaphoreConcurrentLinkedDequeManagedConnectionPool.java

```diff
@@ -573,7 +573,14 @@ else if (debug)
          {
             if (pool.getInternalStatistics().isEnabled())
                pool.getInternalStatistics().deltaBlockingFailureCount();
-
+            synchronized (cls) {
```

Comment:
I have a doubt here: iiuc if we still have some connections in the cls map but the race condition occurred anyway this if will fail and leak would occurr anyway.

Reply:
I agree that is not perfect solution, ideally we would have to have synchronized returnConnection and getConnection but that would be performance issue.  When there are more connections in the cls the probability of the issue is much lower and the consequences are smaller. In our case (pool with size 1) the returned connection is always the same which  is possibly being retrieved from the pool at the same time. Under such circumstances it happens I would say 1 on 10 activemq restarts. So if we have standard configuratiion for example 20 connections in the pool the issue is unlikelly to happen because connections would be rather different. Event if it would happen the consequence is just one less permit to be acquired. In our case it means zero permits and we cannot connect at all. 
Of course I would really appraciate any other solution that would solved the issue.

Reply:
What's more I cannot just release the lock in case of timeout, because I don't know if the lock is properly acquired, do I ? In case of cls.size = 0 I am sure the lock invalid because there is no connection which should keep the lock.

Reply:
@gazda1983 tbh, I would prefer to create a general solution for this issue, even if it implies some performance penalty. I have opened https://github.com/ironjacamar/ironjacamar/pull/792 which adds a lock for idle connection removal. Would you be able to check if it helps for your case? Alternatively, can you share a reproducer so that I can check it myself?